backend/src/test/controller/UserController.test.ts

```diff
@@ -0,0 +1,398 @@
+  });
+
+  test('GET /users', async (done) => {
+    const expectedUser: User = new UserFactory().generate();
+    await userRepository.save(expectedUser);
+    request(app)
+      .get('/users')
+      .expect(200)
+      .end((err, res) => {
+        if (err) return done(err);
+        const users = res.body;
+        expect(users).toHaveLength(1);
+        compareUsers(users[0], expectedUser);
+        done();
+      });
+  });
+
+  test('GET /users/:id', async (done) => {
+    const expectedUser: User = new UserFactory().generate();
+    await userRepository.save(expectedUser);
+    request(app)
+      .get(`/users/${expectedUser.id}`)
+      .expect(200)
+      .end((err, res) => {
+        if (err) return done(err);
+        const user = res.body;
+        compareUsers(user, expectedUser);
+        done();
+      });
+  });
+
+  test('POST /users', async (done) => {
+    const expectedUser: User = new UserFactory().generate();
+    request(app)
+      .post('/users')
+      .send(expectedUser)
+      .expect(201)
+      .end((err, res) => {
+        if (err) return done(err);
+        const user = res.body;
+        compareUsers(user, expectedUser);
+        done();
+      });
+  });
+
+  test('POST /users/ - invalid request body should be caught', async (done) => {
+    const expectedUser: User = new UserFactory().generate();
+    request(app)
+      .post('/users')
+      .send({
+        ...expectedUser,
+        id: undefined, // POST request to users should not contain id
+        email: 'invalid email',
+      })
+      .expect(400)
+      .end((err, res) => {
+        const frontEndErrorObject = res.body;
+        expect(frontEndErrorObject?.errorCode).toEqual('ValidationError');
+        expect(frontEndErrorObject.message).toHaveLength(1);
+        expect(frontEndErrorObject.message[0]).toEqual(
+          '"email" must be a valid email'
+        );
+        done();
+      });
+  });
+
+  test('PATCH /users/:id', async (done) => {
+    const changedProperties = {
+      firstName: 'Diff firstName',
+      lastName: 'Diff lastName',
+      email: 'Diffemail@gmail.com',
+    };
+    const expectedUser: User = new UserFactory().generate();
+    await userRepository.save(expectedUser);
+    request(app)
+      .patch(`/users/${expectedUser.id}`)
+      .send({
+        ...changedProperties,
+      })
+      .expect(204)
+      .end(() => {
+        // check properties of user have changed
+        return getManager().transaction(
+          'READ UNCOMMITTED',
+          async (transactionalEntityManager) => {
+            const userRepository = transactionalEntityManager.getCustomRepository(
+              UserRepository
+            );
+            const user = await userRepository.findOneOrFail({
+              id: expectedUser.id,
+            });
+            expect(user).toMatchObject(changedProperties);
+            done();
+          }
+        );
+      });
+  });
```

Comment:
Ok so I think this will solve one of the issues where we were reading outdated data.
The problem was we issue PATCH request. PATCH request starts a transaction and makes the changes. We call `userRepository.find` and the data hasn't changed. Then the transaction from the PATCH request commits.  Hence we were reading not up to date data.

I think we can fix this by making sure that any repository interaction after issuing supertest request is under a transaction that has isolation level `READ UNCOMMITTED`. So we do not need to depend on the PATCH request to commit.

We still have the problem about the deadlocks, can't seem to figure that out since it occurs very rarely. But I do think we need to prepare our code for deadlocks (e.g. if a deadlock occurs, rollback the current transaction and try again after X seconds)

Reply:
@andrewzulaybar are you okay with me merging still even though sometimes we may face the deadlock issue here? https://github.com/ubclaunchpad/foodies/pull/94#issuecomment-752329183. I think I fixed the other ones

Reply:
Weird I can't reproduce the deadlock anymore....i've tried about 30 times locally LOL

Reply:
I'll proceed with merging. If we keep encountering this problem a lot of times we can bug it and dedicate some further time to look into it!